File: SM3.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define ROTL(x, n) (((x) << (n)) | ((x) >> (32 - (n))))

// SM3初始值
const uint32_t IV[8] = {
    0x7380166F, 0x4914B2B9, 0x172442D7, 0xDA8A0600,
    0xA96F30BC, 0x163138AA, 0xE38DEE4D, 0xB0FB0E4E
};

// 布尔函数
#define FF0(x, y, z) ((x) ^ (y) ^ (z))
#define FF1(x, y, z) (((x) & (y)) | ((x) & (z)) | ((y) & (z)))
#define GG0(x, y, z) ((x) ^ (y) ^ (z))
#define GG1(x, y, z) (((x) & (y)) | ((~(x)) & (z)))

// 置换函数
#define P0(x) ((x) ^ ROTL((x), 9) ^ ROTL((x), 17))
#define P1(x) ((x) ^ ROTL((x), 15) ^ ROTL((x), 23))

// 常量Tj
static const uint32_t T[64] = {
    0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519,
    0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519,
    0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519,
    0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A
};

// 消息填充
void sm3_pad(const uint8_t *msg, size_t len, uint8_t *padded_msg, size_t *padded_len) {
    size_t bit_len = len * 8;
    size_t pad_len = (len % 64 < 56) ? (56 - len % 64) : (120 - len % 64);
    
    memcpy(padded_msg, msg, len);
    padded_msg[len] = 0x80;
    memset(padded_msg + len + 1, 0, pad_len - 1);
    
    for (int i = 0; i < 8; i++) {
        padded_msg[len + pad_len + i] = (bit_len >> (56 - 8 * i)) & 0xFF;
    }
    
    *padded_len = len + pad_len + 8;
}

// 压缩函数
void sm3_compress(uint32_t state[8], const uint8_t block[64]) {
    uint32_t W[68];
    uint32_t W1[64];
    uint32_t A, B, C, D, E, F, G, H;
    uint32_t SS1, SS2, TT1, TT2;
    
    // 消息扩展
    for (int i = 0; i < 16; i++) {
        W[i] = ((uint32_t)block[i * 4] << 24) |
               ((uint32_t)block[i * 4 + 1] << 16) |
               ((uint32_t)block[i * 4 + 2] << 8) |
               ((uint32_t)block[i * 4 + 3]);
    }
    
    for (int j = 16; j < 68; j++) {
        W[j] = P1(W[j-16] ^ W[j-9] ^ ROTL(W[j-3], 15)) ^ ROTL(W[j-13], 7) ^ W[j-6];
    }
    
    for (int j = 0; j < 64; j++) {
        W1[j] = W[j] ^ W[j+4];
    }
    
    // 压缩
    A = state[0]; B = state[1]; C = state[2]; D = state[3];
    E = state[4]; F = state[5]; G = state[6]; H = state[7];
    
    for (int j = 0; j < 64; j++) {
        SS1 = ROTL((ROTL(A, 12) + E + ROTL(T[j], j)), 7);
        SS2 = SS1 ^ ROTL(A, 12);
        
        if (j < 16) {
            TT1 = FF0(A, B, C) + D + SS2 + W1[j];
            TT2 = GG0(E, F, G) + H + SS1 + W[j];
        } else {
            TT1 = FF1(A, B, C) + D + SS2 + W1[j];
            TT2 = GG1(E, F, G) + H + SS1 + W[j];
        }
        
        D = C;
        C = ROTL(B, 9);
        B = A;
        A = TT1;
        H = G;
        G = ROTL(F, 19);
        F = E;
        E = P0(TT2);
    }
    
    state[0] ^= A; state[1] ^= B; state[2] ^= C; state[3] ^= D;
    state[4] ^= E; state[5] ^= F; state[6] ^= G; state[7] ^= H;
}

// SM3哈希函数
void sm3_hash(const uint8_t *msg, size_t len, uint8_t digest[32]) {
    uint8_t *padded_msg = (uint8_t *)malloc(64 * ((len + 72) / 64));
    size_t padded_len;
    uint32_t state[8];
    
    sm3_pad(msg, len, padded_msg, &padded_len);
    
    memcpy(state, IV, sizeof(IV));
    
    for (size_t i = 0; i < padded_len; i += 64) {
        sm3_compress(state, padded_msg + i);
    }
    
    for (int i = 0; i < 8; i++) {
        digest[i * 4] = (state[i] >> 24) & 0xFF;
        digest[i * 4 + 1] = (state[i] >> 16) & 0xFF;
        digest[i * 4 + 2] = (state[i] >> 8) & 0xFF;
        digest[i * 4 + 3] = state[i] & 0xFF;
    }

    free(padded_msg);
}

int main() {
    const char *msg = "Hello,world!";
    uint8_t digest[32];
    
    sm3_hash((const uint8_t *)msg, strlen(msg), digest);
    
    printf("SM3 hash of \"%s\":\n", msg);
    for (int i = 0; i < 32; i++) {
        printf("%02x", digest[i]);
    }
    printf("\n");
    
    return 0;
}

File: sm3.py
# SM3

from __future__ import annotations
from typing import Iterable, Tuple

IV = (
    0x7380166F, 0x4914B2B9, 0x172442D7, 0xDA8A0600,
    0xA96F30BC, 0x163138AA, 0xE38DEE4D, 0xB0FB0E4E
)

T0 = 0x79CC4519
T1 = 0x7A879D8A

def _rotl(x: int, n: int) -> int:
    return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF

def _ff(x: int, y: int, z: int, j: int) -> int:
    if 0 <= j <= 15:
        return x ^ y ^ z
    return (x & y) | (x & z) | (y & z)

def _gg(x: int, y: int, z: int, j: int) -> int:
    if 0 <= j <= 15:
        return x ^ y ^ z
    return (x & y) | (~x & z)

def _p0(x: int) -> int:
    return x ^ _rotl(x, 9) ^ _rotl(x, 17)

def _p1(x: int) -> int:
    return x ^ _rotl(x, 15) ^ _rotl(x, 23)

def _pad(msg_len_bytes: int) -> bytes:
    bit_len = (msg_len_bytes * 8) & ((1 << 64) - 1)
    pad = b'\x80'
    k = (56 - (msg_len_bytes + 1) % 64) % 64
    pad += b'\x00' * k
    pad += bit_len.to_bytes(8, 'big')
    return pad

def _expand(B: bytes) -> Tuple[list, list]:
    assert len(B) == 64
    W = [0]*68
    Wp = [0]*64
    for i in range(16):
        W[i] = int.from_bytes(B[4*i:4*i+4], 'big')
    for j in range(16, 68):
        x = W[j-16] ^ W[j-9] ^ _rotl(W[j-3], 15)
        W[j] = (_p1(x) ^ _rotl(W[j-13], 7) ^ W[j-6]) & 0xFFFFFFFF
    for j in range(64):
        Wp[j] = (W[j] ^ W[j+4]) & 0xFFFFFFFF
    return W, Wp

def compress(V: Tuple[int, ...], B: bytes) -> Tuple[int, ...]:
    A, Bv, C, D, E, F, G, H = V
    W, Wp = _expand(B)
    for j in range(64):
        Tj = T0 if j <= 15 else T1
        SS1 = _rotl(((_rotl(A, 12) + E + _rotl(Tj, j % 32)) & 0xFFFFFFFF), 7)
        SS2 = SS1 ^ _rotl(A, 12)
        TT1 = (_ff(A, Bv, C, j) + D + SS2 + Wp[j]) & 0xFFFFFFFF
        TT2 = (_gg(E, F, G, j) + H + SS1 + W[j]) & 0xFFFFFFFF
        D = C
        C = _rotl(Bv, 9)
        Bv = A
        A = TT1
        H = G
        G = _rotl(F, 19)
        F = E
        E = _p0(TT2)
    return (
        A ^ V[0], Bv ^ V[1], C ^ V[2], D ^ V[3],
        E ^ V[4], F ^ V[5], G ^ V[6], H ^ V[7]
    )

class Hasher:
    __slots__ = ("_state", "_buf", "_length")
    def __init__(self, data: bytes | None = None):
        self._state = IV
        self._buf = bytearray()
        self._length = 0
        if data:
            self.update(data)
    def update(self, data: bytes) -> 'Hasher':
        if not data:
            return self
        self._length += len(data)
        self._buf.extend(data)
        while len(self._buf) >= 64:
            block = bytes(self._buf[:64])
            self._buf = self._buf[64:]
            self._state = compress(self._state, block)
        return self
    def copy(self) -> 'Hasher':
        h = Hasher()
        h._state = tuple(self._state)
        h._buf = bytearray(self._buf)
        h._length = self._length
        return h
    def digest(self) -> bytes:
        tmp = self._buf + _pad(self._length)
        st = self._state
        for i in range(0, len(tmp), 64):
            st = compress(st, tmp[i:i+64])
        return b''.join(x.to_bytes(4, 'big') for x in st)
    def hexdigest(self) -> str:
        return self.digest().hex()
    @classmethod
    def from_state(cls, state_words: Iterable[int], total_len_bytes: int, tail: bytes=b"") -> 'Hasher':
        h = Hasher()
        h._state = tuple(int(x) & 0xFFFFFFFF for x in state_words)
        h._buf = bytearray(tail)
        h._length = total_len_bytes
        return h

def sm3(data: bytes) -> bytes:
    return Hasher(data).digest()

def sm3_hex(data: bytes) -> str:
    return sm3(data).hex()